File: frontend/src/types/listing.ts
export type ListingCondition =
  | ""
  | "New"
  | "Like New"
  | "Used"
  | "For parts/not working";

export type ListingResolvedCondition = Exclude<ListingCondition, "">;

export type UploadedImage = {
  id: string;
  file: File;
  previewUrl: string;
};

export type ListingFormValues = {
  condition: ListingCondition;
  knownIssues: string;
  includedAccessories: string;
  desiredPrice: string;
  sellerNotes: string;
};

export type ItemSpecific = {
  name: string;
  value: string;
};

export type PriceSuggestion = {
  amount: string;
  currency: string;
  confidence: string;
  rationale: string;
};

export type BuyerQuestion = {
  question: string;
  answer: string;
};

export type PublishStatus = "draft" | "published" | "publish_failed";

export type ListingDraft = {
  draftId: string;
  detectedItem: string;
  confidence: string;
  title: string;
  subtitle: string;
  categorySuggestion: string;
  categoryText: string;
  categoryId: string | null;
  condition: ListingResolvedCondition;
  conditionDescription: string;
  description: string;
  itemSpecifics: ItemSpecific[];
  price: string;
  currency: string;
  quantity: number;
  priceSuggestion: PriceSuggestion;
  shippingNotes: string[];
  searchKeywords: string[];
  buyerQuestions: BuyerQuestion[];
  missingInfoWarnings: string[];
  imageUrls: string[];
  merchantLocationKey: string | null;
  paymentPolicyId: string | null;
  fulfillmentPolicyId: string | null;
  returnPolicyId: string | null;
  publishStatus: PublishStatus;
  sku: string | null;
  offerId: string | null;
  listingId: string | null;
  listingUrl: string | null;
};

export type DraftUpdatePayload = {
  title: string;
  categoryText: string;
  condition: ListingResolvedCondition;
  description: string;
  itemSpecifics: ItemSpecific[];
  price: string;
  quantity: number;
  priceSuggestion: {
    rationale: string;
  };
};

File: backend/app/models/listing.py
from typing import Literal

from pydantic import BaseModel, Field


ConditionValue = Literal["New", "Like New", "Used", "For parts/not working"]
PublishStatusValue = Literal["draft", "published", "publish_failed"]


class ItemSpecific(BaseModel):
    name: str
    value: str


class PriceSuggestion(BaseModel):
    amount: str
    currency: str = "CAD"
    confidence: str
    rationale: str


class BuyerQuestion(BaseModel):
    question: str
    answer: str


class ListingDraft(BaseModel):
    draftId: str
    detectedItem: str
    confidence: str
    title: str
    subtitle: str
    categorySuggestion: str
    categoryText: str
    categoryId: str | None = None
    condition: ConditionValue
    conditionDescription: str
    description: str
    itemSpecifics: list[ItemSpecific]
    price: str
    currency: str = "CAD"
    quantity: int = 1
    priceSuggestion: PriceSuggestion
    shippingNotes: list[str]
    searchKeywords: list[str]
    buyerQuestions: list[BuyerQuestion]
    missingInfoWarnings: list[str]
    imageUrls: list[str] = Field(default_factory=list)
    merchantLocationKey: str | None = None
    paymentPolicyId: str | None = None
    fulfillmentPolicyId: str | None = None
    returnPolicyId: str | None = None
    publishStatus: PublishStatusValue = "draft"
    sku: str | None = None
    offerId: str | None = None
    listingId: str | None = None
    listingUrl: str | None = None
